src/libcrun/cgroup-resources.c

```diff
@@ -711,12 +715,31 @@ write_devices_resources_v2 (int dirfd, runtime_spec_schema_defs_linux_device_cgr
         {
           crun_error_release (err);
           *err = tmp_err;
-          return ret;
+          return -1;
         }
       if (rootless)
-        can_skip = true;
+        *can_skip = true;
     }
 
+  return 0;
+}
+
+static int
+write_devices_resources_v2 (int dirfd, runtime_spec_schema_defs_linux_device_cgroup **devs, size_t devs_len,
+                            libcrun_error_t *err)
+{
+  int ret;
+  bool can_skip;
+
+  ret = write_devices_resources_v2_internal (dirfd, devs, devs_len, err);
+  if (LIKELY (ret == 0))
+    return 0;
```

Comment:
`crun_error_release (err);` is missing

Reply:
Hmm, I'm not sure I follow. When `write_devices_resources_v2_internal` returns `0` there's no `err` to release.

Reply:
Ah, you're right. Sorry. I already forgot that `can_skip_devices()` contains   `crun_error_release()`.
`can_skip_devices()` expects the input argument `libcrun_error_t *err` to represent an already created error.

Reply:
After
```
  ret = can_skip_devices (&can_skip, devs, devs_len, err);
  if (UNLIKELY (ret < 0))
    return ret;
```

https://github.com/kolyshkin/crun/blob/af2b3a8161979b2ec25ceb2d2dfd28e557c62ed3/src/libcrun/cgroup-resources.c#L739C1-L741C16

we know that `ret` is `0` because ` can_skip_devices()` only returns `-1` or `0`

To me this code

```
   if (can_skip)
     {
       crun_error_release (err);
       ret = 0;
     }
```
https://github.com/kolyshkin/crun/blob/af2b3a8161979b2ec25ceb2d2dfd28e557c62ed3/src/libcrun/cgroup-resources.c#L743-L747

is a bit confusing because `ret` is already known to be `0` and why is the error only released when `can_skip` is true.

Reply:
@eriksjolund you're right, this needs to be straightened out one way or another. I will open a followup PR.